README.md
JEvents
=======

Main JEvents Repository for core component, modules and plugins.

The master branch is currently for JEvents 3.4 which works with Joomla 3.4+
JEvents 3.2 is latest branch for Joomla 2.5 but there will be no new features added here - only major bug fixes.
JEvents 3.1 is the old JEvents 3.1.x series which are no longer developed.





CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

#### 10-07-2018 - Tony Partridge
 * Added language strings for Full Calendar Theme
 * added {{DESCRIPTION_ADDSLASHES}} shortcode 
 
#### 24-04-2018 - Tony Partridge
### Fixed
 * Accessibility improvements to the category select and pagination limit list.
 
#### 02-03-2018 - Tony Partridge
### Fixed
 * event list being filtered after cancelling event creation i.e. events filtered by published state.

#### 27-02-2018 - Tony Partridge
### Fixed
 * Saving bug resulting in call on null / rp_id() / location()
 * Event and event repeat selector not working due to incorrect JSession Check
 * Emptying trashed events
 * Joomla 3.6.x fatal error when trying to load Joomla 3.7 new calendar. 
 * List of Events view return blank on Repeat Summary
 * Set calendar popup width to be 300px; a nice size rather than 100%.
 
#### 17-02-2018 Geraint Edwards
 * JEvents 3.4.44
 * Authorisation check on save should check creator and editor permissions
 * Fix for colour picker being compromised by Joomla settings
 * Allow jomsocial group events plugin to be picked up in edit page layout editor
 * No-op code added to Category filter that can be used by an override to force category filters to apply to ALL selected filters and ANY of the selected filters
 * Workaround for lack of support for electric calendar in recent Joomla date picker update
 * Category selector replaced with style category name if only one choice is available
 * Allow date formatting in layout editor to use date() function syntax as well as strftime format
 * Replace deprecated Joomla method calls
 * Improved meta tag setting 
 * PHP 7.2 Support Changes
 * Fix for setting published state in iCal and csv import
 * Update code to current standards in installer
 * Add default all day event tick selection
 * fixed nonendtime default selection.
 * Fix noendtime config option, was not working. 
 * column selector should show textarea to allow advanced configuration
 * Fix for default category setting when changing calendar in event creation
 * Fix for timezone problems when editing repeats of existing events
 * Add ical and edit dialogs as options for calendar cell
 * calendar module navigation needs to handle UTF-8 data
 * csv export should allow all events to be output even if block show all is enabled
 * iCal export wasn't using maxYear and minYear so DST transitions were not always output
 * gwejson library needs to add slashes to the message
 * EXT Theme Fix Legend to show categories names always.
 * Ext theme legend width to auto so nothing is cut off.
 * Fix for default category setting when creating new event
 * Fix guest check, WE only want to show the message IF NOT guest.
 * Fix for missing cal_day declaration in default layout
 * Add additional param to getListQuery to allow setting if list or item
 * Added filtering to iCalRepeat to be same as iCalEvent

#### 29-08-2017 Geraint Edwards
 * JEvents 3.4.43
 * New fix for #389 Fix for Authorised Users not being able to edit own events when they hit the event limit
 * Stop check on freq==none being case sensitive

#### 26-08-2017 Geraint Edwards
 * JEvents 3.4.42
 * Reverse Tony's changes on max event creation check for authorised users since it broke Joomla ACL permissions
 * Fix for timezone specified repeating events which were not adjusted correctly when re-editing an event

#### 22-08-2017 Geraint Edwards
 * JEvents 3.4.41
 * revert changes to JOIN queries on rrule which had exposed a problem with iCal imported events not showing in the frontend.

#### 22-08-2017 Geraint Edwards
 * JEvents 3.4.40
 * Move Timezone input to before date/time on event edit page (where enabled)
 * Add loc_id integer field to database for locations to improve performance of DB queries on large sites
 * Improve irregular date imports
 * Add location index to event detail table for performance gains
 * Fix toolbar buttons in backend
 * Use Inner join instead of left join for performance gains
 * stray use of | instead of $separator
 * Allow contact field to be truncated in latest events module
 * no end time should default to 0 not ""
 * Allow CDATA through template replacer - was replacing closing ]] with }} which was not good for Google 
 * Fix for Authorised Users not being able to edit own events when they hit the event limit.
 * Added where clause to smart search plugin to only find published events.
 * Modify input sanitization where the user has allowed raw.
 * Make sure visibile filters check only is applied when using JEvents filter module or modules with filters in their params

#### 01-07-2017 Tony Partridge
* Updated Changelog
* Added new default option for no specific endtime.
* Disable system messages for iCal Reloading when a guest.

#### 25-06-2016 Tony Partridge
* Added new custom layout tag: {{{Past or Future:PAST_OR_FUTURE}} to output text past or future depending if the event is in the past or future. Handy for CSS Classing.

#### 21-06-2017 Tony Partridge
* Added CREATED: to the iCal export events following iCal Specification.

#### 20-06-2017 Tony Partridge
* Update ext modstyle.css to stop centralising latest events caused by  my change on the 15th.

#### 16-06-2017 Tony Partridge
* Set admin emails to run through the defaultloadedtemplate.php parser. This means tags used in the Event Details custom layout can now be used within the admin emails.

#### 15-05-2017 Tony Partridge
 * JEvents 3.4.39
 * W3C Validation fix in ext latest events module (Removed inline tags)
 * Added version compare to filtering method onSave.
 * Fixed {DURATION} output which broke in 3.4.38 by Tony's (int) conversion.
 
#### 11-05-2017 Tony Partridge
* JEvents 3.4.38 
* Fixed issue with inputFilter saving on pre J3.7.1

#### 10-05-2017 Geraint Edwards
* JEvents 3.4.37
* For for checkbox and radio box required fields check

package/install.php
<?php

/**
 * JEvents Component for Joomla 2.5.x
 *
 * @version     JEVENTS_VERSION
 * @releasedate JEVENTS_DATE
 * @package     JEvents
 * @copyright   Copyright (C) 2008-JEVENTS_COPYRIGHT GWESystems Ltd, 2006-2008 JEvents Project Group
 * @license     GNU/GPLv2, see http://www.gnu.org/licenses/gpl-2.0.html
 * @link        http://www.jevents.net
 */

// no direct access
defined('_JEXEC') or die('Restricted access');

use Joomla\CMS\Language\Text;
use Joomla\CMS\Filesystem\File;
use Joomla\CMS\Factory;

jimport('joomla.filesystem.folder');
jimport('joomla.filesystem.file');
jimport('joomla.application.component.helper');

class Pkg_JeventsInstallerScript
{
	public function preflight($type, $parent)
	{

		define('JEVENTS_MINIMUM_PHP', '7.0.0');

		if (version_compare(PHP_VERSION, JEVENTS_MINIMUM_PHP, '<'))
		{
			Factory::getApplication()->enqueueMessage(Text::sprintf("COM_JEVENTS_PHP_VERSION_WARNING", PHP_VERSION), 'warning');
		}

		// Joomla! broke the update call, so we have to create a workaround check.
		$db = Factory::getDbo();
		$db->setQuery("SELECT enabled FROM #__extensions WHERE element = 'com_jevents'");
		$is_enabled = $db->loadResult();

		if (!$is_enabled)
		{
			$this->hasJEventsInst = 0;
			if (version_compare(JVERSION, '3.8.0', '<'))
			{
				Factory::getApplication()->enqueueMessage('Warning! You are running a very insecure version of Joomla! <br/>Please update Joomla! to at least 3.8.0 before installing JEvents. This will also prevent issues with JEvents', 'warning');

				return false;
			}

			return;
		}
		else
		{
			$this->hasJEventsInst = 1;
			if (version_compare(JVERSION, '3.8.0', '<'))
			{
				Factory::getApplication()->enqueueMessage('This version of JEvents is designed for Joomla 3.8.0 and later.<br/>Please update Joomla! before upgrading JEvents to this version', 'warning');

				return false;
			}

			return;
		}
	}

	public function update($parent)
	{

		$this->postflightHandler("update", $parent);

		return true;
	}

	public function postflightHandler($type, $parent)
	{

		$app    = Factory::getApplication();

		// CSS Styling:
		?>
		<style type="text/css">
			.adminform tr th:first-child {
				display: none;
			}

			table.adminform tr td {
				padding: 15px;
			}

			div.jev_install {
				background-color: #f4f4f4;
				border: 1px solid #ccc;
				border-radius: 5px;
				padding: 10px;
			}

			.installed {
				clear: both;
				display: inline-block;
			}

			.installed ul {
				width: 350px;
				padding-left: 0px;
				border: 1px solid #ccc;
				border-radius: 5px;
			}

			.installed ul li:first-child {
				border-top-left-radius: 5px;
				border-top-right-radius: 5px;
			}

			.installed ul li:last-child {
				border-bottom-left-radius: 5px;
				border-bottom-right-radius: 5px;
			}

			.installed ul li {
				padding: 8px;
				list-style-type: none;
			}

			.installed ul li:nth-child(odd) {
				background-color: #fff;
			}

			.installed ul li:nth-child(even) {
				background-color: #D6D6D6;
			}

			.proceed {
				display: inline-block;
				vertical-align: top;
			}

			div.proceed ul {
				text-align: center;
				list-style-type: none;
			}

			div.proceed ul li {
				padding: 5px;
				background-color: #fff;
				border: 1px solid #ccc;
				margin-bottom: 10px;
				border-radius: 5px;
			}
		</style>
		<?php
		// End of CSS Styling
		if ($this->hasJEventsInst == 1)
		{
			$inst_text = Text::_('JEV_INST_VERSION_UPRG');
			$logo      = "JeventsTransparent3.png";
		}
		else
		{
			$inst_text = Text::_('JEV_INST_VERSION');
			$logo      = "JeventsTransparent.png";
		}

		// TODO Replace ADD_VERSION
		echo "<div class='jev_install'>
				<div class='jev_logo'><a href='index.php?option=com_jevents' ><img src='https://www.jevents.net/logo/$logo' /></a></div>
				<div class='version'><h2 class='gsl-h2'>" . $inst_text . ": ADD_VERSION </h2></div>
				<h3>Exciting New Styling & Joomla 4 Support</h3>
				<p>JEvents 3.6 brings you</p>
				<ul>
					<li>A new backend theme (with new configuration options)</li>
					<li>A new dashboard that gives you an overview and graphical summary of your events (you can choose this or the events list as your landing page)</li>
					<li>A new, more modern, interface for event creation/editing (with new configuration options)</li>
					<li>Joomla 4.0 support</li>
				</ul>
				<p>Please see our latest <a href='https://www.jevents.net/blog/jevents-3-6-exciting-new-styling-featured' target='_blank'>blog post</a> for more information about these features and how they can be configured.</p>
				<div class='installed'>
					<ul>
						<li><a href='index.php?option=com_jevents' >JEvents Core Component</a></li>
						<li>JEvents Module - Latest Events </li>
						<li>JEvents Module - Mini Calendar </li>
						<li>JEvents Module - Filter </li>
						<li>JEvents Module - Legend </li>
						<li>JEvents Module - Switch View </li>
						<li>JEvents Plugin - Search </li>
						<li>JEvents Plugin - Finder </li>
					</ul>
				</div>

				<div class='proceed'>
					<ul>
						<li><a href='index.php?option=com_jevents&task=params.edit' alt='JEvents Configuration'><img src='components/com_jevents/assets/images/jevents_config_sml.png' alt='Configuration Page' /><br/> Configuration</a><br/></li>
						<li><a href='https://www.jevents.net/forum' alt='JEvents Forum'><img src='components/com_jevents/assets/images/support_forum.jpg' alt='JEvents Forum' /><br/>Support Forums</a><br/></li>
						<li><a href='https://www.jevents.net/docs/jevents' alt='JEvents Documentation'><img src='components/com_jevents/assets/images/documentation.jpg' alt='JEvents Documentation' /><br/>Documentation</a></li>
					</ul>
				</div>";

		if ($this->hasJEventsInst == 0)
		{
			// enable plugin
			$db    = Factory::getDbo();
			$query = "SELECT * FROM #__extensions WHERE name='plg_content_finder' and type='plugin' and element='finder'";
			$db->setQuery($query);
			$finder_q = $db->loadObject();
			$finder   = $finder_q->enabled;

			$query = 'SHOW TABLES LIKE "' . $db->getPrefix() . 'finder_types"';
			$db->setQuery($query);
			$finder_types = $db->loadObjectList();

			if (!count($finder_types))
			{
				echo "<div class='alert alert-warning'> Warning! your Joomla! installation is missing the finder_types database table.<br/><br/> You should run a database check and then fix if an error is reporting by <a href='index.php?option=com_installer&view=database' class='btn-warning btn button'>Clicking Here</a></div>";
			}

			if ($finder == 1 && count($finder_types))
			{
				$query = "UPDATE #__extensions SET enabled=1 WHERE folder='finder' and type='plugin' and element='jevents'";
				$db->setQuery($query);
				$db->execute();
			}

			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='search' and type='plugin' and element='eventsearch'";
			$db->setQuery($query);
			$db->execute();

			// Enable new JEvents Plugin
			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='content' and type='plugin' and element='jevents'";
			$db->setQuery($query);
			$db->execute();

			// Enable JSON Plugin
			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='system' and type='plugin' and element='gwejson'";
			$db->setQuery($query);
			$db->execute();

			// Enable Jevents Installer Plugin
			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='installer' and type='plugin' and element='jeventsinstaller'";
			$db->setQuery($query);
			$db->execute();

		}
		else
		{
			jimport('joomla.filesystem.file');
			// Ok Flatplus clean up to remove helpers
			$file1 = JPATH_SITE . '/components/com_jevents/views/flatplus/helpers/flatplusloadedfromtemplate.php';
			$file2 = JPATH_SITE . '/components/com_jevents/views/flatplus/helpers/flatpluseventmanagementdialog.php';
			$file3 = JPATH_SITE . '/components/com_jevents/views/flatplus/helpers/flatplusicaldialog.php';

			if (File::exists($file1)) File::delete($file1);
			if (File::exists($file2)) File::delete($file2);
			if (File::exists($file3)) File::delete($file3);

			$file4 = JPATH_SITE . '/components/com_jevents/libraries/checkconflict.php';
			if (File::exists($file4)) File::delete($file4);

			// Lets make sure our Core plugin is enabled..
			$db    = Factory::getDbo();
			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='content' and type='plugin' and element='jevents'";
			$db->setQuery($query);
			$db->execute();

			// Enable JSON Plugin
			$query = "UPDATE #__extensions SET enabled=1 WHERE folder='system' and type='plugin' and element='gwejson'";
			$db->setQuery($query);
			$db->execute();

		}

		echo "</div>";
		// Joomla updater special case
		if ($app->input->getCmd("option") == "com_installer" && $app->input->getCmd("view") == "update")
		{
			$app->enqueueMessage("<div class='jev_logo'><img src='https://www.jevents.net/logo/JeventsTransparent3.png' /></div>" . Text::_('JEV_INST_VERSION_UPRG') . " :: ADD_VERSION", 'message');
		}

	}

	public function install($parent)
	{

		$this->postflightHandler("update", $parent);

		return true;
	}

	/*
	 * enable the plugins
	 */

	public function uninstall($parent)
	{

		$uninstall_text  = Text::_('JEV_SORRY_THAT_YOU_UNINSTALL');
		$uninstall_text2 = Text::_('JEV_PLEASE_LET_US_KNOW_WHY');
		if ($uninstall_text == 'JEV_SORRY_THAT_YOU_UNINSTALL')
		{
			$uninstall_text  = "We are sorry that you have uninstalled JEvents";
			$uninstall_text2 = "Please let us know why at our <a href='https://www.jevents.net/forum'>support forum</a>  so we can improve our product offering for future users.";
		}
		echo "<div class='jev_install'>
				<div class='jev_logo'><img src='https://www.jevents.net/logo/JeventsTransparent2.png' /></div>
				<div class='version'><h2 class='gsl-h2'>" . $uninstall_text . "</h2></div>
				<div class='installed'>
					<h4>" . $uninstall_text2 . "</h4>
                                        <br/><br/><br/>
				</div>";

		return true;
	}

	/*
 * enable the plugins
 */

	public function postflight($type, $parent)
	{

		return;
		//return $this->postflightHandler($type, $parent);
	}

}

